### ISSUE TITLE
cmd/link: openshift build fails due to another problem with too far branches on ppc64le

### ISSUE THREAD

**Issue by laboger** [golang/go#20497]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go 1.9

### What operating system and processor architecture are you using (`go env`)?
Ubuntu 16.04

### What did you do?
Built upstream openshift binaries, which builds using packages from Kubernetes.

### What did you expect to see?
Binaries built that work.

### What did you see instead?
Some binaries now fail due to a new linker limit being exceeded, which causes the GNU linker to generate plt stubs for long calls.  The stub generated in this case by the GNU linker expects that R2 contains the address of the TOC (according to the PPC64LE v2 ABI) but code generated by golang in this case does not maintain that value in R2.  As a result the generated address is bad and attempts to branch to a bad address.

I know trying to set up and maintain R2 in this case is not the best solution.  I believe the solution as requested here https://github.com/golang/go/issues/17917 is the right answer but I need to test it out more.  (I had some concerns about this solution which is why I wasn't able to get it into go 1.9).

To reproduce this, clone the repo from here:
https://github.com/openshift/origin.git
cd origin; make all 

Try to run one of the tests in:
_output/local/bin/linux/ppc64le with the -h option and the result is a SEGV.

**Comment by laboger** [comment]
The failure occurs the first time main.init tries to call a plt_branch sequence like this:

000000001380448c <00000038.plt_branch.36:b9ef>:
    1380448c:   ff ff 82 3d     addis   r12,r2,-1
    13804490:   10 4e 8c e9     ld      r12,19984(r12)
    13804494:   a6 03 89 7d     mtctr   r12
    13804498:   20 04 80 4e     bctr

But even though this is a dynamic executable built by the GNU linker, the code is not generated to load R2 or maintain it.  As a result, this sequence expects R2 to be valid when it is not, and jumps off to a bad address, or 0.

This is the first time I've seen a plt_branch to handle this situation, if the branch offset is not too large then it is just a long_branch and that works fine.

I am able to build valid binaries successfully by using -buildmode=pie.

In reading the PPC64 v2 ABI, it sounds like the linker should know whether or not the procedure has initialized R2 and generate the call stub based on that information?  If that is true then it shouldn't be creating this type of call stub unless it doesn't have correct information about the procedure making the call.  I'm trying to track that down.

**Comment by jcajka** [comment]
I have hit this too(Fedora27/26, go1.8.3), with bins(origin/genman,origin/openshift) ~300M+ in size and was able to get (at least partially) functional binaries using linkmode internal(for non shared bins, haven't tried shared at all), but I haven't verified that binaries are generally correct(yet). They don't segfault when executed(which  seems to be according to your notes).

**Comment by laboger** [comment]
The workaround for this is to use -bulidmode=pie when building the executables.  I've tested with it and seems to work fine.  It will ensure that r2 is valid as needed.

A better fix is planned for Go 1.9 and is being tested to verify no regressions are introduced.

**Comment by jcajka** [comment]
Actually internal linker is not a workaround(at least not in full static Go/default build, surprisingly it is when you have intermediate files left from failed build with external linker, haven't investigated further this path yet), it fails with relocation overflow. With build mode pie(without -linkshared) and go1.8.3 I'm now hitting illegal instruction in genman, investigating further.

seems as jump to ctr which point in to data

  >│0x242e1ec0                      .long 0x0                                                                                                      
          
#0  0x00000000242e1ec0 in ?? () at /usr/lib/golang/src/runtime/asm_ppc64x.s:109
#1  0x00000000242e1db4 in runtime.rt0_go () at /usr/lib/golang/src/runtime/asm_ppc64x.s:17
#2  0x00003fffb7d2291c in generic_start_main (main=0x242e49b0 <main>, argc=<optimized out>, argv=0x3fffffffeed8, auxvec=0x3ffffffff040, init=<optimized out>, rtld_fini=<optimized out>, 
    stack_end=<optimized out>, fini=<optimized out>) at ../csu/libc-start.c:289
#3  0x00003fffb7d22b18 in __libc_start_main (argc=<optimized out>, argv=<optimized out>, ev=<optimized out>, auxvec=<optimized out>, rtld_fini=<optimized out>, stinfo=<optimized out>, 
    stack_on_entry=<optimized out>) at ../sysdeps/unix/sysv/linux/powerpc/libc-start.c:102
#4  0x0000000000000000 in ?? ()

**Comment by laboger** [comment]
Can you provide directions on how you build genman?  This was reported to me when building openshift using the directions above, and I don't see how to build genman.  If I build openshift with -buildmode=pie using go 1.8.3 it works AFAICT, and it certainly doesn't die in __libc_start_main.

Maybe you noticed, the upstream patch was submitted this morning, and we are testing the go 1.8.3 patch now so that should be available very soon.

**Comment by laboger** [comment]
I will be changing my solution for this problem as a result of the discussion in #20492.

**Comment by jcajka** [comment]
@laboger genman is built using hack/generate-docs.sh, I did some tracing and it seem that following will build genman(invoked in tree of failed build of OS).

```
export GOPATH=origin/_output/local/go
go build -x -ldflags='-linkmode=internal' -pkgdir origin/_output/local/pkgdir/linux/ppc64le -o genman -tags 'include_gcs include_oss containers_image_openpgp ' github.com/openshift/origin/tools/genman

```
(path adjustment needed)
Recently I have done most of the investigation by hacking common.sh like this way(which is our proposed workaround)

```
+        linkshared=""
+      if [[ "${platform}" == "linux/ppc64le" ]]; then
+        local_ldflags+=" -buildmode=pie"
+        linkshared="-linkshared"
+      fi
+
       if [[ ${#nonstatics[@]} -gt 0 ]]; then
-        GOOS=${platform%/*} GOARCH=${platform##*/} go install \
+        GOOS=${platform%/*} GOARCH=${platform##*/} go install $linkshared \
           -pkgdir "${OS_OUTPUT_PKGDIR}/${platform}" \
           -tags "${OS_GOFLAGS_TAGS-} ${!platform_gotags_envvar:-}" \
           -ldflags="${local_ldflags}" \

```

**Comment by laboger** [comment]
A few notes on the workaround.  Most of my testing had been done upstream, and I made the wrong assumption that go 1.8 would behave the same.  But I've since discovered an upstream fix (not in go 1.8)  that is needed to make this workaround work in all cases.  Sorry about that confusion. @jcajka
